tests/protocol/test_wire.py

```diff
@@ -516,227 +521,235 @@ def test_get_ext_conf_with_uri(self, *args):
+                self.assertTrue(http_request.urls[2].endswith('/machine/?comp=goalstate'), "The host channel should have been refreshed the goal state")
+                self.assertTrue(http_request.urls[3].endswith('/extensionArtifact'), "The third attempt should have been over the host channel")
+                self.assertEquals(HostPluginProtocol.is_default_channel(), False, "The host channel should not have been set as the default")
+
+    def test_fetch_manifest_should_not_invoke_host_channel_when_direct_channel_succeeds(self):
+        manifest_url = 'https://fake_host/fake_manifest.xml'
+        manifest_xml = '<?xml version="1.0" encoding="utf-8"?><PluginVersionManifest/>'
 
-                        self.assertEquals(HostPluginProtocol.is_default_channel(), True)
+        with mock_wire_protocol.create(mockwiredata.DATA_FILE) as protocol:
+            original_http_request = restutil.http_request
 
-        # Reset default channel
-        HostPluginProtocol.set_default_channel(False)
+            def http_request(method, url, *args, **kwargs):
+                if method == 'GET':
+                    if url == manifest_url:
+                        return MockResponse(body=manifest_xml.encode('utf-8'), status_code=200)
+                    elif url.endswith('/extensionArtifact'):
+                        self.fail('The Host GA Plugin should not have been invoked')
+                return original_http_request(method, url, *args, **kwargs)
 
-    @skip_if_predicate_true(lambda: True, "Needs to be re-enabled before release 2.2.47")
-    @patch("azurelinuxagent.common.protocol.wire.WireClient.get_goal_state")
-    @patch("azurelinuxagent.common.protocol.hostplugin.HostPluginProtocol.get_artifact_request")
-    def test_fetch_manifest_should_retry_the_host_channel_after_refreshing_the_host_plugin(self,
-                                                                                           mock_get_artifact_request,
-                                                                                           *args):
-        mock_get_artifact_request.return_value = "dummy_url", "dummy_header"
-        client = WireClient("foo.bar")
+            with patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request) as mock_request:
+                HostPluginProtocol.set_default_channel(False)
 
-        HostPluginProtocol.set_default_channel(False)
+                manifest = protocol.client.fetch_manifest([VMAgentManifestUri(uri=manifest_url)])
 
-        mock_failed_response = MockResponse(body=b"", status_code=httpclient.GONE)
-        mock_successful_response = MockResponse(body=b"OK", status_code=200)
+                self.assertEquals(manifest, manifest_xml)
+                self.assertTrue(any(args[0][0] == 'GET' and args[0][1] == manifest_url for args in mock_request.call_args_list))  # direct channel
```

Comment:
Don't we expect the first GET to succeed since we're mocking it? Why are we expecting several requests?

Reply:
Note that this test follows a different pattern than the other ones. This one was one of the first test I started with.

The underlying implementation of the protocol can trigger multiple requests while executing those fetch* and upload* methods. For example, if we get to use the host plugin we will trigger a request for the goal state, followed by puts to the health service for the plugin, etc.

The mocks in the other tests keep track of the interesting requests in http_request.urls and discard the rest. In this one test I am using mock_request.call_args_list instead. If you run the test under the debugger and set a breakpoint on the first assert you can check the call list and you will see a request for the goal state on top of the the GET for the manifest.

I think the pattern I used on the other tests allows for stronger assertions, and also makes tests a lot more readable so I'll change this one to follow that pattern as well.

Also, post-release, I'll see if I can capture that pattern in a more general tool that we can use for this kind of tests.

Reply:
I captured the pattern I was mentioning as mock_http_request in mocks.py

Reply:
Got it, thanks for the update.